sc-memory/sc-core/sc-store/sc-container/sc-dictionary/sc_dictionary.c

```diff
@@ -0,0 +1,381 @@
+  while (i < size)
+  {
+    sc_uint8 num;
+    sc_char_to_sc_int(**sc_string_ptr, &num, &node->mask);
+    // define prefix
+    if (SC_DICTIONARY_NODE_IS_NOT_VALID(node->next[num]))
+    {
+      node->next[num] = _sc_dictionary_node_initialize();
+
+      sc_dictionary_node * temp = node->next[num];
+
+      temp->offset = *sc_string_ptr;
+      temp->offset_size = size - i;
+
+      node = temp;
+
+      break;
+    }
+    // visit next substring
+    else if (node->next[num]->offset != null_ptr)
+    {
+      sc_dictionary_node * moving = node->next[num];
+
+      sc_uint32 j = 0;
+      for (; i < size && j < moving->offset_size && tolower(moving->offset[j]) == tolower(**sc_string_ptr);
+           ++i, ++j, ++*sc_string_ptr)
+        ;
+
+      if (j != moving->offset_size)
+      {
+        saved_offset_size = moving->offset_size;
+
+        node->next[num] = _sc_dictionary_node_initialize();
+
+        sc_dictionary_node * temp = node->next[num];
+
+        temp->offset = moving->offset;
+        temp->offset_size = j;
+      }
+
+      node = node->next[num];
+
+      // insert intermediate node for prefix end branching
+      if (j < moving->offset_size)
+      {
+        sc_char * offset_ptr = &*(node->offset + j);
+
+        sc_char_to_sc_int(*offset_ptr, &num, &node->mask);
+        if (node->next[num] == null_ptr)
+        {
+          node->next[num] = moving;
+
+          sc_dictionary_node * temp = node->next[num];
+
+          temp->offset = offset_ptr;
+          temp->offset_size = saved_offset_size - j;
+        }
+      }
+    }
+    else
+    {
+      node = node->next[num];
+      ++*sc_string_ptr;
+      ++i;
+    }
+  }
+
+  g_free(sc_string_ptr);
+  return node;
+}
+
+sc_dictionary_node * sc_dictionary_append(sc_dictionary * dictionary, sc_char * sc_string, sc_uint32 size, void * value)
+{
+  sc_dictionary_node * node = sc_dictionary_append_to_node(dictionary->root, sc_string, size);
+
+  if (node->data_list == null_ptr)
+    sc_list_init(&node->data_list);
+
+  sc_list_push_back(node->data_list, value);
+  return node;
+}
+
+sc_dictionary_node * sc_dictionary_remove_from_node(
+    sc_dictionary_node * node,
+    const sc_char * sc_string,
+    sc_uint32 index)
+{
+  // check prefixes matching
+  sc_uint32 string_size = strlen(sc_string);
+  sc_uint8 num = 0;
+  sc_char_to_sc_int(sc_string[index], &num, &node->mask);
+  if (num != 0 && SC_DICTIONARY_NODE_IS_VALID(node->next[num]))
+  {
+    node->next[num] = sc_dictionary_remove_from_node(node->next[num], sc_string, index + node->next[num]->offset_size);
+
+    if (SC_DICTIONARY_NODE_IS_VALID(node->next[num]))
+      return node;
+  }
+
+  // check suffixes matching
+  if (index == string_size)
+  {
+    sc_char * str = g_new0(sc_char, node->offset_size + 1);
+    memcpy(str, node->offset, node->offset_size);
+
+    if (strcmp(str, sc_string + (string_size - node->offset_size)) == 0)
+    {
+      g_free(str);
+      return null_ptr;
+    }
+    g_free(str);
+  }
```

Comment:
These lines are also used in another function. Can this be moved to a separate function?